### PR TITLE
Refactor page code

### PR DESCRIPTION
**Pull request by Morganamilo**
Refactors all of the message decoding logic into one trait with associated types. This simplifies the type signatures all the way down the page types.

This also adds a do nothing decode implementer so that any type can easily be passed and paged.

### CODE REVIEW

File: src/rest.rs [hunk 1]
```diff
@@ -924,22 +923,40 @@ impl Format {
...
-    fn handle(&self, msg: &mut T) {
-        msg.decode(self.opts.as_ref());
+impl Decode for Stats {
+    type Data = ();
+    type Item = Self;
+    fn decode(_item: &mut Self::Item, _data: &Self::Data) {}
+}
+
+impl Decode for PresenceMessage {
+    type Data = Option<ChannelOptions>;
```
Comment (lmars) [inline src/rest.rs:950]: The associated type is named `Data`, then here we set it to `ChannelOptions`, and around the code we call it `handler`, which I found a little confusing.

Perhaps we should name the associated type `Options` (i.e. options for performing the decode), and then have:

```rust
type Options = Option<ChannelOptions>;
```

and:

```rust
paginated_request_with_options(... opts: T::Options)
```

?

File: src/rest.rs [hunk 2]
```diff
@@ -424,11 +438,11 @@ impl<'a> Channel<'a> {
     ///
     /// Returns a history::RequestBuilder which is used to set parameters
     /// before sending the history request.
-    pub fn history(&self) -> history::PaginatedRequestBuilder<Message> {
-        self.rest.paginated_request(
+    pub fn history(&self) -> PaginatedRequestBuilder<Message> {
+        self.rest.paginated_request_with_handler(
```
Comment (lmars) [inline src/rest.rs:442]: It seems like every call to `self.rest.paginated_request` has become `self.rest.paginated_request_with_handler` in this PR, so perhaps just change the signature of `self.rest.paginated_request`?

File: src/rest.rs [hunk 3]
```diff
@@ -924,22 +923,40 @@ impl Format {
...
+impl Decode for Message {
+    type Data = Option<ChannelOptions>;
+    type Item = Self;
+
+    fn decode(item: &mut Self::Item, data: &Self::Data) {
+        crate::rest::decode(&mut item.data, &mut item.encoding, data.as_ref());
     }
 }
 
-impl<T: Decode> http::PaginatedItemHandler<T> for MessageItemHandler {
```
[the remark this answers is not in the document]
Comment (Morganamilo) [reply src/rest.rs:941]: Probably! This work was split over 3 months so I may have missed some bits...

File: src/rest.rs (hunk 1 above)
Comment (Morganamilo) [reply src/rest.rs:950]: Options may confuse with std::option::Option<T>. The name could be different though I guess, and handler changed to data/whatever we change data to yeah.

File: src/rest.rs (hunk 2 above)
Comment (Morganamilo) [reply src/rest.rs:442]: paginated_request_with_handler is the special function that lets you do pass in some extra data, while paginated_request is more basic but probably want most of the time.

Internally paginated_request_with_handler is used more as messages require the decode step. But these functions meant to satisfy the request() spec point. So even though the other function isn't used by use, it's meant to be used by the user,

Also one is implemented in terms of the other, there's no code duplication.

File: src/history.rs [hunk 4]
```diff
@@ -1,10 +1 @@
-use crate::{http, rest};
```
[the remark this answers is not in the document]
Comment (Morganamilo) [reply src/history.rs:1]: Yep, another case of 3 month break in work lol.

File: src/rest.rs (hunk 1 above)
Comment (Morganamilo) [reply src/rest.rs:950]: Changed handler to data for consistency. When we decide on a name we can then change all data to whatever.

Comment (Morganamilo) [reply src/rest.rs]: Couldn't think of a better name so went with Options

File: src/rest.rs [hunk 5]
```diff
@@ -198,13 +204,21 @@ impl Rest {
...
     /// response is unsuccessful (i.e. the status code is not in the 200-299
     /// range).
-    pub fn paginated_request<T: http::PaginatedItem, U: http::PaginatedItemHandler<T>>(
-        &self,
+    pub fn paginated_request_with_data<'a, T: Decode + 'a>(
+        &'a self,
+        method: http::Method,
+        path: &str,
+        data: T::Options,
+    ) -> http::PaginatedRequestBuilder<T> {
```
Comment (lmars) [inline src/rest.rs:212]:
```suggestion
    pub fn paginated_request_with_options<'a, T: Decode + 'a>(
        &'a self,
        method: http::Method,
        path: &str,
        options: T::Options,
    ) -> http::PaginatedRequestBuilder<T> {
```

**Review by lmars — APPROVED** [review]
LGTM (apart from suggested function rename), much nicer!

**Review by lmars — APPROVED** [review]
🙌

**Pull request merged**